workflow-service/src/main/java/com/redhat/parodos/workflow/WorkFlowDelegate.java

```diff
@@ -43,34 +45,57 @@ public WorkFlowDelegate(BeanWorkFlowRegistryImpl beanWorkFlowRegistry) {
 					workFlowRequestDTO.getWorkFlowName(), WorkContextDelegate.Resource.ARGUMENTS,
 					WorkFlowDTOUtil.convertArgumentListToMap(workFlowRequestDTO.getArguments()));
 		}
-		if (workFlowRequestDTO.getWorks() != null)
-			workFlowRequestDTO.getWorks()
-					.forEach(work -> initWorkContext(workContext, work, workFlowRequestDTO.getWorkFlowName()));
+		if (masterWorkFlowDefinitionDto.getWorks() != null && !masterWorkFlowDefinitionDto.getWorks().isEmpty())
+			masterWorkFlowDefinitionDto.getWorks()
+					.forEach(work -> initWorkContext(workContext,
+							findWorkFromRequest(workFlowRequestDTO, work.getName()), work,
+							masterWorkFlowDefinitionDto.getName()));
 		return workContext;
 	}
 
 	private void initWorkContext(WorkContext workContext, WorkFlowRequestDTO.WorkRequestDTO workRequestDTO,
-			String parentWorkflowName) {
+			WorkDefinitionResponseDTO workDefinitionResponseDTO, String parentWorkflowName) {
 		WorkContextDelegate.write(workContext, WorkContextDelegate.ProcessType.WORKFLOW_EXECUTION,
-				workRequestDTO.getWorkName(), WorkContextDelegate.Resource.PARENT_WORKFLOW, parentWorkflowName);
-		if (workRequestDTO.getArguments() != null && !workRequestDTO.getArguments().isEmpty()) {
+				workDefinitionResponseDTO.getName(), WorkContextDelegate.Resource.PARENT_WORKFLOW, parentWorkflowName);
+
+		if (workRequestDTO != null && workRequestDTO.getArguments() != null
+				&& !workRequestDTO.getArguments().isEmpty()) {
 			WorkContextDelegate.write(workContext,
-					WorkType.WORKFLOW.name().equals(workRequestDTO.getType())
+					WorkType.WORKFLOW.name().equalsIgnoreCase(workRequestDTO.getType())
 							? WorkContextDelegate.ProcessType.WORKFLOW_EXECUTION
 							: WorkContextDelegate.ProcessType.WORKFLOW_TASK_EXECUTION,
 					workRequestDTO.getWorkName(), WorkContextDelegate.Resource.ARGUMENTS,
 					WorkFlowDTOUtil.convertArgumentListToMap(workRequestDTO.getArguments()));
 		}
-		if (workRequestDTO.getWorks() != null)
-			workRequestDTO.getWorks().forEach(work -> initWorkContext(workContext, work, workRequestDTO.getWorkName()));
+		if (workDefinitionResponseDTO.getWorks() != null)
+			workDefinitionResponseDTO.getWorks().forEach(work -> initWorkContext(workContext,
+					findSubWorkFromRequest(workRequestDTO, work.getName()), work, workDefinitionResponseDTO.getName()));
+	}
+
+	private WorkFlowRequestDTO.WorkRequestDTO findWorkFromRequest(WorkFlowRequestDTO workFlowRequestDTO, String name) {
+		if (workFlowRequestDTO != null && workFlowRequestDTO.getWorks() != null)
+			for (WorkFlowRequestDTO.WorkRequestDTO workRequestDTO : workFlowRequestDTO.getWorks())
+				if (name.equals(workRequestDTO.getWorkName()))
+					return workRequestDTO;
+		return null;
+	}
+
+	private WorkFlowRequestDTO.WorkRequestDTO findSubWorkFromRequest(WorkFlowRequestDTO.WorkRequestDTO workRequestDTO,
+			String name) {
+		if (workRequestDTO != null && workRequestDTO.getWorks() != null)
```

Comment:
I would change this to make it simple, like 

if (workRequestDTO == null || workRequestDTO.getWorks() == null) return null

It's more cognitive based